eva/storage/video_storage_engine.py

```diff
@@ -0,0 +1,108 @@
+from eva.readers.opencv_reader import OpenCVReader
+from eva.models.storage.batch import Batch
+from eva.configuration.configuration_manager import ConfigurationManager
+from eva.utils.logging_manager import LoggingManager
+from eva.utils.logging_manager import LoggingLevel
+
+
+class VideoStorageEngine(AbstractStorageEngine):
+
+    def __init__(self):
+        self.metadata = 'metadata'
+        self.upload_path = Path(
+            ConfigurationManager().get_value('storage', 'path_prefix'))
+        self.curr_version = ConfigurationManager().get_value(
+            'storage', 'video_engine_version')
+
+    def create(self, table: DataFrameMetadata, video_file: Path):
+        # Create directory to store video and metadata related to the video
+        dir_path = Path(table.file_url)
+        try:
+            Path.mkdir(dir_path, parents=True)
+        except FileExistsError:
+            LoggingManager().log('Failed to load the video as directory \
+                                already exists {}'.format(
+                dir_path), LoggingLevel.ERROR)
+        found = False
+        # copy video from the provided path to eva datasets
+        if Path(video_file).exists():
+            shutil.copy2(video_file, str(dir_path))
+            found = True
+        # check in the upload directory
+        else:
+            video_path = Path(self.upload_path / video_file)
+            if video_path.exists():
+                shutil.copy2(str(video_path), str(dir_path))
+                found = True
+        if not found:
+            LoggingManager().log('Failed to find the video file {}'.format(
+                video_file), LoggingLevel.ERROR)
+            return False
+        self._create_video_metadata(dir_path, video_file.name)
+        return found
+
+    def write(self, table: DataFrameMetadata, rows: Batch):
+        pass
+
+    def read(self,
+             table: DataFrameMetadata,
+             batch_mem_size: int,
+             predicate_func=None) -> Iterator[Batch]:
+
+        metadata_file = Path(table.file_url) / self.metadata
+        video_file_name = self._get_video_file_path(metadata_file)
+        video_file = Path(table.file_url) / video_file_name
+        reader = OpenCVReader(str(video_file), batch_mem_size=batch_mem_size)
+        for batch in reader.read():
+            yield batch
+
+    def _get_video_file_path(self, metadata_file):
+        with open(metadata_file, 'rb') as f:
+            version = int.from_bytes(f.read(4), byteorder='big')
+            if version > self.curr_version:
+                LoggingManager().log('Invalid metadata version {}'
+                                     .format(version), LoggingLevel.ERROR)
+                return False
+            length = int.from_bytes(f.read(4), byteorder='big')
+            path = f.read(length)
+            return Path(path.decode())
+
+    def _create_video_metadata(self, dir_path, video_file):
```

Comment:
What is the use for the metadata?

Reply:
Added this so that in the future if we plan to add support for splitting the video into smaller chunks or even tiling video we can store that info in the metadata